rmp-serde/src/encode.rs

```diff
@@ -285,111 +280,258 @@ impl<'a, W: VariantWriter> serde::Serializer for Serializer<'a, W> {
+    }
+
+    #[inline]
+    fn serialize_map(&mut self, len: Option<usize>) -> Result<Self::MapState, Self::Error> {
+        let len = match len {
             Some(len) => len,
             None => return Err(Error::UnknownLength),
         };
 
         try!(write_map_len(&mut self.wr, len as u32));
 
-        while let Some(()) = try!(depth_count!(self.depth, visitor.visit(self))) { }
-
         Ok(())
     }
 
-    fn serialize_map_elt<K, V>(&mut self, key: K, value: V) -> Result<(), Error>
-        where K: serde::Serialize,
-              V: serde::Serialize,
+    #[inline]
+    fn serialize_map_key<T>(&mut self,
+                            _state: &mut Self::MapState,
+                            key: T)
+                            -> Result<Self::MapState, Self::Error>
+        where T: serde::Serialize
+    {
+        key.serialize(self)
+    }
+
+    #[inline]
+    fn serialize_map_value<T>(&mut self,
+                              _state: &mut Self::MapState,
+                              value: T)
+                              -> Result<Self::MapState, Self::Error>
+        where T: serde::Serialize
     {
-        try!(key.serialize(self));
         value.serialize(self)
     }
 
+    #[inline]
+    fn serialize_map_end(&mut self, _state: Self::MapState) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    #[inline]
     fn serialize_unit_struct(&mut self, _name: &'static str) -> Result<(), Error> {
         try!(self.vw.write_struct_len(&mut self.wr, 0));
 
         Ok(())
     }
 
-    fn serialize_struct<V>(&mut self, _name: &str, mut visitor: V) -> Result<(), Error>
-        where V: serde::ser::MapVisitor,
+    #[inline]
+    fn serialize_newtype_struct<T>(&mut self,
+                                   _name: &'static str,
+                                   value: T)
+                                   -> Result<(), Self::Error>
+        where T: serde::Serialize
     {
-        let len = match visitor.len() {
-            Some(len) => len,
-            None => return Err(Error::UnknownLength),
-        };
-
-        try!(self.vw.write_struct_len(&mut self.wr, len as u32));
+        value.serialize(self)
+    }
 
-        while let Some(()) = try!(depth_count!(self.depth, visitor.visit(self))) { }
+    #[inline]
+    fn serialize_newtype_variant<T>(&mut self,
+                                    name: &'static str,
+                                    variant_index: usize,
+                                    variant: &'static str,
+                                    _value: T)
+                                    -> Result<(), Self::Error>
+        where T: serde::Serialize
+    {
+        self.serialize_tuple_variant(name, variant_index, variant, 1)
+    }
 
-        Ok(())
+    #[inline]
+    fn serialize_struct(&mut self,
+                        _name: &'static str,
+                        len: usize)
+                        -> Result<Self::StructState, Self::Error> {
+        self.serialize_map(Some(len))
```

Comment:
Does this mean that the struct with always be written as a Map and we can no longer write out as an Array type? Trying to serialize a struct with a StructArrayWriter still outputs a Map it seems because the VariantWriter is ignored now.